common/journald/src/stream.rs

```diff
@@ -0,0 +1,344 @@
+use systemd::journal::{Journal, JournalFiles, JournalRecord, JournalSeek};
+use chrono::{Local, TimeZone};
+use http::types::body::LineBuilder;
+use metrics::Metrics;
+use log::{warn};
+use futures::stream::Stream as FutureStream;
+use std::{
+    mem::drop,
+    path::PathBuf,
+    pin::Pin,
+    sync::{
+        mpsc::{sync_channel, Receiver, TryRecvError},
+        Arc,
+        Mutex,
+    },
+    task::{Context, Poll, Waker},
+    thread::{self, JoinHandle},
+    time::UNIX_EPOCH,
+};
+
+const KEY_TRANSPORT: &str = "_TRANSPORT";
+const KEY_HOSTNAME: &str = "_HOSTNAME";
+const KEY_COMM: &str = "_COMM";
+const KEY_PID: &str = "_PID";
+const KEY_MESSAGE: &str = "MESSAGE";
+
+const TRANSPORT_AUDIT: &str = "audit";
+const TRANSPORT_DRIVER: &str = "driver";
+const TRANSPORT_SYSLOG: &str = "syslog";
+const TRANSPORT_JOURNAL: &str = "journal";
+const TRANSPORT_STDOUT: &str = "stdout";
+const TRANSPORT_KERNEL: &str = "kernel";
+
+#[derive(Clone)]
+pub enum Path {
+    Directory(PathBuf),
+    Files(Vec<PathBuf>),
+}
+
+enum RecordStatus {
+    Line(LineBuilder),
+    BadLine,
+    NoLines,
+}
+
+struct SharedState {
```

Comment:
I've since added a second field to the object to ensure the thread is cleaned up when the main stream is dropped. I don't think AtomicCell is the correct primitive to be using here and I don't think the single use case warrants importing yet another package (although with sccache now I can hardly tell the difference :P)